### DISCUSSION TITLE
pad test unexpected output

### DISCUSSION THREAD

**Discussion by GrumpyGopher** [darthcloud/BlueRetro#962] [Q&A] [answered]
I'm working on a modified version of the pad test with the intent of outputting the data to a serial port of an Arduino board.  I was able to get the data to Arduino in a working manner but noticed a couple things that I wanted to ask about.

When connecting an 8bitdo SN30 pro+ viewing the output from pad test in the console is as expected.  When the state of the controller changes a message is outputted.  If I'm not doing anything on the controller then no messages are received.

If I connect an 8bitdo Neo Geo Pad or PS4, or PS5 pad the console is flooded with messages regardless of weather anything is being pressed on the controller.  I was able to work around this for Neo Geo Pad since it is all digital by just verifying that the state has changed in some way before outputting the message but was wondering if there is a way to prevent this behavior and only get debug messages when controller state has changed?

The other thing I noticed was an inconsistency on what is reported from the analog sticks.  Depending on the controller the ranges where different:
8bitdo SN30 Pro and Pro+ had a range of -32768 to 32767
PS4 and PS5 controllers had a range of -127 to 127
Xbox One controllers had and range slightly higher than the 8bitdo controllers

This makes it a little more difficult to write something on the Arduino that will work with all controllers.  Any suggestions on how to get this data more consistent?

Thanks

**Comment by darthcloud** [comment, marked as answer]
Yes by default it give you the original range of the controller.

You could force it to output the converted value for a specific system with modifying the to the pad_test config:

Replace:
CONFIG_BLUERETRO_ADAPTER_INPUT_DBG=y

With:
CONFIG_BLUERETRO_ADAPTER_INPUT_MAP_DBG=y

And replace:
CONFIG_BLUERETRO_SYSTEM_PARALLEL_1P=y

With a system that got dual joystick and trigger like GameCube.

**Reply by GrumpyGopher** [reply]
That makes sense I didn't even think of that.  Adding CONFIG_BLUERETRO_ADAPTER_INPUT_MAP_DBG=y and selecting Dreamcast as my system did limit the amounts but it seems like with this enabled I'm only getting debug information from the first controller where as with the other way I was getting debug information from all 6 connected controllers.

**Reply by GrumpyGopher** [reply]
I see the problem by default the wired message is always output information for index 0 
adapter_debug_wired_print(&ctrl_output[0]);
Once I updated this with the index of the controller I'm getting all the info now.

Thanks for all the help this will work perfectly!

**Reply by darthcloud** [reply]
Ok feel free to submit PR, sound like a bug

**Comment by darthcloud** [comment]
This output report data as it is received.

Some controller like 8bitdo and Xbox update on changes.

Other like PlayStation and Nintendo updates on a interval regardless of change.

As it is there is no way to filter the message. I guess you could modify the code to do so easily.

**Comment by darthcloud** [comment]
You might wan to add to config:

CONFIG_BLUERETRO_JSON_DBG=y

This will output json formatted string that are easy to parse with cjson.

you can use idf.py menuconfig to edit the config, check the blueretro menu
BUt editing the config itself is easy enough too.
